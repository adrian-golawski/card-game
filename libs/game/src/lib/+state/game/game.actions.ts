import { createAction, props } from '@ngrx/store';

export const startNewGame = createAction('[Game] Start New Game');
export const startNewGameSuccess = createAction(
  '[Game] Start New Game Success'
);
export const startNewGameFailure = createAction(
  '[Game] Start New Game Failure',
  props<{ error: Error }>()
);
