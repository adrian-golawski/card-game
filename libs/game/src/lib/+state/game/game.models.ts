export interface GameEntity {
  id: number;
}
