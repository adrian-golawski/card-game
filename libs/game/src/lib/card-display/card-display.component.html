<div class="playing-card">
  <div [ngClass]="{ reversed: this.reversed }" class="flip-card">
    <div
      [style.background-image]="
        card && card.image
          ? 'url(' + card.image + ')'
          : 'linear-gradient(white, black)'
      "
      class="card-front"
    ></div>
    <div class="card-back"></div>
  </div>
</div>
