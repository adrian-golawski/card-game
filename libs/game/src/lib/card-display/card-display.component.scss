.playing-card {
  height: 100%;
  width: 100%;

  .flip-card {
    position: relative;

    width: 100%;
    height: 100%;

    transition: transform 1s;
    transform-style: preserve-3d;

    &.reversed {
      transition: transform 1s;
      transform-style: preserve-3d;
      transform: rotateY(180deg);
    }
  }

  img {
    width: 100%;
    height: 100%;
  }

  .card-back,
  .card-front {
    position: absolute;

    width: 100%;
    height: 100%;

    border-radius: 6px;
    box-shadow: rgba(0, 0, 0, 0.3) 2px 2px 5px;

    backface-visibility: hidden;
  }

  .card-front {
    background-size: cover;
    background-repeat: no-repeat;
    background-position: center center;
  }

  .card-back {
    transform: rotateY(-180deg);

    background: radial-gradient(
        circle,
        transparent 20%,
        #f7f7f7 20%,
        #f7f7f7 80%,
        transparent 80%,
        transparent
      ),
      radial-gradient(
          circle,
          transparent 20%,
          #f7f7f7 20%,
          #f7f7f7 80%,
          transparent 80%,
          transparent
        )
        20px 20px,
      linear-gradient(#d40000 4px, transparent 4px) 0 -2px,
      linear-gradient(90deg, #d40000 4px, transparent 4px) -2px 0;
    background-color: rgba(0, 0, 0, 0);
    background-size: auto, auto, auto, auto;
    background-color: #f7f7f7;
    background-size: 40px 40px, 40px 40px, 20px 20px, 20px 20px;

    border: 1px solid #d40000;
  }
}
