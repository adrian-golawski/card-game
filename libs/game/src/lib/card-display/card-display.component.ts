import { Component, Input } from '@angular/core';
import { Card } from '@card-game/cards';

@Component({
  selector: 'card-game-card-display',
  templateUrl: './card-display.component.html',
  styleUrls: ['./card-display.component.scss'],
})
export class CardDisplayComponent {
  @Input() reversed: boolean;
  @Input() card?: Card;

  constructor() {}
}
