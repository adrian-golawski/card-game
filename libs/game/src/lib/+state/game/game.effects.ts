import { Injectable } from '@angular/core';
import * as CardActions from '@card-game/cards';
import { Actions, createEffect, ofType } from '@ngrx/effects';
import { Action } from '@ngrx/store';

import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';

import * as GameActions from './game.actions';

@Injectable()
export class GameEffects {
  constructor(private readonly actions$: Actions) {}

  public startNewGame$: Observable<Action> = createEffect(() =>
    this.actions$.pipe(
      ofType(GameActions.startNewGame),
      map(CardActions.getNewDeck)
    )
  );
}
