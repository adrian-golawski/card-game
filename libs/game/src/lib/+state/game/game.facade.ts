import { Injectable } from '@angular/core';
import { select, Store } from '@ngrx/store';

import { Observable } from 'rxjs';

import * as GameActions from './game.actions';
import * as fromGame from './game.reducer';
import * as GameSelectors from './game.selectors';

@Injectable()
export class GameFacade {
  score$: Observable<number> = this.store.pipe(select(GameSelectors.getScore));
  gameInProgress$: Observable<boolean> = this.store.pipe(
    select(GameSelectors.gameInProgress)
  );
  roundsLeft$: Observable<number> = this.store.pipe(
    select(GameSelectors.roundsLeft)
  );
  betGiven$: Observable<boolean> = this.store.pipe(
    select(GameSelectors.betGiven)
  );
  betLower$: Observable<boolean> = this.store.pipe(
    select(GameSelectors.betLower)
  );
  gameLoading$: Observable<boolean> = this.store.pipe(
    select(GameSelectors.gameLoading)
  );

  constructor(private readonly store: Store<fromGame.GamePartialState>) {}

  startNewGame(): void {
    this.store.dispatch(GameActions.startNewGame());
  }

  giveBet(lower: boolean): void {
    this.store.dispatch(GameActions.betGiven({ lower }));
  }
}
