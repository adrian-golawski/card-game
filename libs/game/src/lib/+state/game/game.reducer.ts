import { Action, createReducer, on } from '@ngrx/store';

import * as GameActions from './game.actions';
export const GAME_FEATURE_KEY = 'game';

export interface State {
  score: number;
  gameActive?: boolean;
  roundsLeft?: number;
}

export interface GamePartialState {
  readonly [GAME_FEATURE_KEY]: State;
}

export const initialState: State = {
  score: 0,
};

const gamesReducer = createReducer(
  initialState,
  on(GameActions.startNewGameSuccess, (state) => ({
    ...state,
    score: 0,
    gameActive: true,
    roundsLeft: 30,
  }))
);

export function reducer(state: State, action: Action): State {
  return gamesReducer(state, action);
}
