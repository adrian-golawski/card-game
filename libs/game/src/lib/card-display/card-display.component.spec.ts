import { ComponentFixture, TestBed } from '@angular/core/testing';
import { CardValue } from '@card-game/cards';

import { Shallow } from 'shallow-render';

import { GameModule } from '../game.module';
import { CardDisplayComponent } from './card-display.component';

describe('CardDisplayComponent', () => {
  let shallow: Shallow<CardDisplayComponent>;

  beforeEach(() => {
    shallow = new Shallow(CardDisplayComponent, GameModule);
  });

  it('should match the snapshot reversed', async () => {
    const { fixture } = await shallow.render({
      bind: { reversed: true },
    });

    expect(fixture).toMatchSnapshot();
  });

  it('should match the snapshot with no card given', async () => {
    const { fixture } = await shallow.render({
      bind: { reversed: false },
    });

    expect(fixture).toMatchSnapshot();
  });

  it('should match the snapshot with card given', async () => {
    const { fixture } = await shallow.render({
      bind: {
        reversed: false,
        card: {
          image: 'https://deckofcardsapi.com/static/img/KH.png',
          value: CardValue.KING,
          suit: 'HEARTS',
          code: 'KH',
        },
      },
    });

    expect(fixture).toMatchSnapshot();
  });
});
