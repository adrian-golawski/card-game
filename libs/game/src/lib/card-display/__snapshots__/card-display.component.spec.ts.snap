// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardDisplayComponent should match the snapshot reversed 1`] = `
<ng-component>
  <card-game-card-display>
    <div
      class="playing-card"
    >
      <div
        class="flip-card reversed"
      >
        <div
          class="card-front"
        />
        <div
          class="card-back"
        />
      </div>
    </div>
  </card-game-card-display>
</ng-component>
`;

exports[`CardDisplayComponent should match the snapshot with card given 1`] = `
<ng-component>
  <card-game-card-display>
    <div
      class="playing-card"
    >
      <div
        class="flip-card"
      >
        <div
          class="card-front"
          style="background-image: url(https://deckofcardsapi.com/static/img/KH.png);"
        />
        <div
          class="card-back"
        />
      </div>
    </div>
  </card-game-card-display>
</ng-component>
`;

exports[`CardDisplayComponent should match the snapshot with no card given 1`] = `
<ng-component>
  <card-game-card-display>
    <div
      class="playing-card"
    >
      <div
        class="flip-card"
      >
        <div
          class="card-front"
        />
        <div
          class="card-back"
        />
      </div>
    </div>
  </card-game-card-display>
</ng-component>
`;
